src/vec_common_ppc.h

```diff
@@ -28,14 +28,173 @@
+ *
+ * Vector constants are often need to for; masking operations,
+ * range checks, permute selection, and radix conversion.
+ * Also compiler support for large integer and floating-point
+ * constants may be limited by the compiler.
+ * For example the GCC compilers support the (vector) __int128 type
+ * but do not directly support __int128 (39 digit) decimal constants.
+ * Another example is __float128 where the type and Q suffix constants
+ * are recent additions.
+ * In both cases we need to construct; large numeric constants,
+ * special values (infinity and NaN), masks for manipulating the sign
+ * bit and exponent bits.
+ * Often these values will be constructed from vectors of word or
+ * doubleword constants.
+ *
+ * \note GCC does not support expressing an integer constant of type
+ * __int128 for targets where long long integer is less than 128
+ * bits wide.  This applies to the PowerPC target as the long long
+ * type is reserved for 64-bit integers.
+ *
+ * \note GCC __float128 support for the PowerPC target began with
+ * GCC 6. In GCC 6 __float128 support is off by default and has to be
+ * explicitly enabled via the '-mfloat128' option. Starting with GCC 7,
+ * __float128 is enabled by default with VSX support.
+ *
+ * Defining large constants for vectors is
+ * complicated by <I>little-endian</I> (LE) support as
+ * specified in the OpenPOWER ABI and as implemented in the compilers.
+ * Little-endian changes the effective vector element numbering and
+ * the order of constant elements in initializers.  But the __int128
+ * numerical order of magnitude or floating-point format does not
+ * change in registers. The high order bits are on the left and the
+ * low order bits are on the right.
+ *
+ * So for example:
+ * \code
+  const vui32_t signmask = { 0x80000000, 0, 0, 0 };
+  const vui32_t expmask = { 0x7fff0000, 0, 0, 0 };
+ * \endcode
+ * are correct sign and exponent masks for __float128 in big
+ * endian (BE) but would be incorrect for little endian (LE).
+ * To get correct results for both endians, one could code something
+ * like this:
+ * \code
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  const vui32_t signmask = { 0, 0, 0, 0x80000000 };
+  const vui32_t expmask = { 0, 0, 0, 0x7fff0000 };
+#else
+  const vui32_t signmask = { 0x80000000, 0, 0, 0 };
+  const vui32_t expmask = { 0x7fff0000, 0, 0, 0 };
+#endif
+ * \endcode
+ * But this gets tedious after the first dozen times.
+ * Also this can be confusing because it does not appear to the match
+ * the floating-point format diagrams in the PowerISA. The sign-bit
+ * and the exponent are always on the left.
+ *
+ * So this header provides endian sensitive macros that maintain
+ * consistent "magnitude" order. For example:
+ * \code
+  const vui32_t signmask = CONST_VINT128_W (0x80000000, 0, 0, 0);
+  const vui32_t expmask = CONST_VINT128_W (0x7fff0000, 0, 0, 0);
+ * \endcode
+ * This is always correct in either endian.
```

Comment:
Is it worth adding something like "this is also valid on platforms (environments?) where long long integers are less than 128 bits (64bit and 32bit environments)"?

Reply:
Yes that is correct. But we don't support 32-bit little endian. Only RHEL7/SLES11 support 32-bit Big Endian where this is not an issues. SLES12, Ubuntu are 64-bit LE only. I suspect RHEL8 will be also. I would rather not address this now and see if anyone shows up and cares. Is that OK?